core/src/main/java/io/questdb/griffin/engine/functions/groupby/LastStringGroupByFunction.java

```diff
@@ -0,0 +1,46 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.groupby;
+
+import io.questdb.cairo.ArrayColumnTypes;
+import io.questdb.cairo.ColumnType;
+import io.questdb.cairo.map.MapValue;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.engine.functions.GroupByFunction;
+import io.questdb.griffin.engine.functions.StrFunction;
+import io.questdb.griffin.engine.functions.UnaryFunction;
+import org.jetbrains.annotations.NotNull;
+
+public class LastStringGroupByFunction extends FirstStringGroupByFunction implements GroupByFunction, UnaryFunction {
+    public LastStringGroupByFunction(@NotNull Function arg) {
+        super(arg);
+    }
+
```

Comment:
toString() creates a lot of unnecessary objects .
It'd be better to store strings as StringSinks and use 
public void getStr(Record rec, CharSink sink) 
instead. 
stringValues list and object it contains could end up taking a lot of memory  or even trigger OOM so it would be good to put a max memory limit .

Reply:
If I do `stringValues.add(this.arg.getStr(record))` directly, then it will always refer to the same value later (for all groups) when querying it, i.e. it will always contain for all groups the first value observed for the last group. I assume this is due to the underlying handling of records where I sadly don't yet understand whats going on.

Reply:
That's why I recommended copying content to stringSink with getStr(Record rec, CharSink sink), e.g. 
`this.arg.getStr(record,` stringValues.get(stringValues.size()-1) ));`

Reply:
I replaced it with an approach that copies the charsequence into a DirectCharSink instead. Hoping that yields better performance.

Reply:
Code below is practically the same as before, it just creates and releases new DirectCharSink instead of String . 
That's going to create tremendous amount of garbage objects on  a big table . 
You can check it yourself if you start qdb with "-verbose:gc" flag and issue query against table with millions of strings .
It'd be better to keep existing object but run clear() on it and then set new content . 

```
    public void computeNext(MapValue mapValue, Record record) {
        CharSequence cs = this.arg.getStr(record);
        DirectCharSink copy = new DirectCharSink(cs.length());
        copy.put(cs);
        int ix = mapValue.getInt(this.valueIndex);
        stringValues.get(ix).close();
        stringValues.set(ix, copy);
    }
```

Reply:
Should be now improved

Reply:
👍